### PR TITLE
Add script to update generated SDS to SCAP 1.3

### PR DESCRIPTION
**Pull request by yuumasato**
#### Description:

- Add script that updates generated DataStream to be compliant with SCAP 1.3.

#### How to test

- Use [scapval-1.3.2](https://csrc.nist.gov/Projects/Security-Content-Automation-Protocol/SCAP-Releases/SCAP-1-3)
  - Use `openjdk-1.8.0` on a RHEL environment. (for some reason it doesn't work on Fedora)
  - Increase allocated memory: `-Xmx1024m`
- Workaround bugs in scap-schematron-rules-1.3.2 (just to test that scapval passes):
  - `group_id` and `user_id` cannot be zero, tracked here https://github.com/OVAL-Community/OVAL/issues/23
    - workaround it : 
      - `sed -i -E "s/(group_id .*\">)/\1-1/" ./ssg-rhel7-ds.xml`
      - `sed -i -E "s/(user_id .*\">)/\1-1/" ./ssg-rhel7-ds.xml`
  - OCIL checks for `security_patches_up_to_date` is causing fail of SRC-377, when requirement is about OVAL checks. I have e-mailed NIST about this to try to sort out what is the issue.
    - workaround it:
      - `awk '/check-content-ref.*up_to_date_ocil/{for(x=NR-1;x<=NR+1;x++)d[x];}{a[NR]=$0}END{for(i=1;i<=NR;i++)if(!(i in d))print a[i]}' ./ssg-rhel7-ds.xml > ./ssg-rhel7-ds-1.3.xml`
- Validate it: `java -Xmx1024m -jar scapval-1.3.2.jar -file ./ssg-rhel7-ds-1.3.xml -scapversion 1.3 > scap-1.3.log`

- Report of scapval with workarounds: https://fedorapeople.org/~wsato/scap-1.3/validation-report.html

#### Rationale:

- SCAP 1.3 should supersede SCAP 1.2 
- In the near future OpenSCAP should be update to handle generation of SCAP 1.3 DS.
  - More specifically the command `oscap ds sds-compose`

### CODE REVIEW

File: build-scripts/update_sds_to_scap_1_3.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+def move_patches_up_to_date_to_source_data_stream_component(datastreamtree):
+    ds_checklists = datastreamtree.find(".//ds:checklists", ns)
+
+    for component_ref in ds_checklists:
+        component_id = component_ref.get('{%s}href' % xlink_ns)
+        component_id = component_id[1:]
+
+        # Locate the <xccdf:check> of the <xccdf:Rule> with id security_patches_up_to_date
+        oval_check = datastreamtree.find(".//ds:component[@id='%s']//xccdf:Rule[@id='xccdf_org.ssgproject.content_rule_security_patches_up_to_date']/xccdf:check[@system='%s']" % (component_id, oval_ns), ns )
```
Comment (jan-cerny) [inline build-scripts/update_sds_to_scap_1_3.py]: This way of using namespaces us not supported in Python 2.6. You need to use eg. 
`find(".//{%s}Benchmark" % (ssg.constants.XCCDF12_NS))`

Comment (jan-cerny) [inline build-scripts/update_sds_to_scap_1_3.py]: Pyhton 2.6 doesn't support XPath attributes, you have to loop and filter manually:
```
for rule in rules:
    id_ = rule.get("id")
    if id_ == "xccdf_org.ssgproject.content_rule_security_patches_up_to_date":
        checks = ....
...
```

File: build-scripts/update_sds_to_scap_1_3.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        if not uris:
+            uri = ssg.xml.ElementTree.Element('{%s}uri' % cat_ns,
+                                    attrib = {
+                                        'name' : component_ref_name,
+                                        'uri' : component_ref_uri })
+            catalog.append(uri)
+
+        # Add the component-ref to list of datastreams' checks
+        ds_checks = datastreamtree.find(".//ds:checks", ns)
+        check_component_ref = ds_checks.findall("ds:component-ref[@id='%s']" % component_ref_uri[1:], ns)
```
Comment (jan-cerny) [inline build-scripts/update_sds_to_scap_1_3.py]: It would be nice to have a comment explaining this `[1:]`.

Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: I have made it explicit, and also added a comment.

File: build-scripts/update_sds_to_scap_1_3.py (hunk 1 above)
Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: I have stopped using the namespace dictionary in favor of suggested namespace approach.

Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: That is unfortunate.
I have implemented the filtering manually.

File: build-scripts/update_sds_to_scap_1_3.py (hunk 2 above)
Comment (jan-cerny) [reply build-scripts/update_sds_to_scap_1_3.py]: Great, but there are still 2 other occurrences of this `[1:]` which might not be clear.

Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: Sure, thank you for asking for these clarifications.
I have added some comments.

File: build-scripts/update_sds_to_scap_1_3.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                break
+        if not uri_exists:
+            uri = ssg.xml.ElementTree.Element('{%s}uri' % cat_ns)
+            uri.set('name', component_ref_name)
+            uri.set('uri', component_ref_uri)
+            catalog.append(uri)
+
+        # The component-ref ID is the catalog uri without leading '#'
+        component_ref_feed_id = component_ref_uri[1:]
+
```
Comment (jan-cerny) [inline build-scripts/update_sds_to_scap_1_3.py]: When I build RHEL 7 data stream I get in scap_org.open-scap_cref_ssg-rhel7-pcidss-xccdf-1.2.xml component-ref this element:
`<ns2:uri name="ecurity-data-oval-com.redhat.rhsa-RHEL7.xml.bz2" uri="#scap_org.open-scap_cref_ecurity-data-oval-com.redhat.rhsa-RHEL7.xml.bz2"/>`. There is missing `s` 2 times.

File: build-scripts/update_sds_to_scap_1_3.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    from urlparse import urlparse
+
+import ssg.constants
+import ssg.xml
+ocil_ns = ssg.constants.ocil_namespace
+
+oval_ns = ssg.constants.oval_namespace
+xccdf_ns = ssg.constants.XCCDF12_NS
+ds_ns = ssg.constants.datastream_namespace
+xlink_ns = ssg.constants.xlink_namespace
```
Comment (jan-cerny) [inline build-scripts/update_sds_to_scap_1_3.py]: This is just renaming variables. Please use them directly.

File: build-scripts/update_sds_to_scap_1_3.py (hunk 3 above)
Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: Nice catch, I ended up introducing a bug during manual filtering.

File: build-scripts/update_sds_to_scap_1_3.py (hunk 4 above)
Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: I have updated the import definition and changed code to use them.

File: build-scripts/update_sds_to_scap_1_3.py (hunk 3 above)
Comment (yuumasato) [reply build-scripts/update_sds_to_scap_1_3.py]: Commits 15c70d0 and 304270f rename variables to make them clear about what components they are about.

**Comment by yuumasato** [comment]
@jan-cerny Thank you for all the reviews.
The content is finally building on RHEL6, and the PEP8 issues are fixed as well.

**Comment by yuumasato** [comment]
Here is validation report from rhel8 content, with `sed` and `awk` workarounds applied.

**Comment by yuumasato** [comment]
With these changes, OpenSCAP 1.3.0 will fail to scan if `--fetch-remote-resources` is not provided.
```
$ oscap-ssh rhel7.6-kicked 22 xccdf eval --profile pci-dss ./ssg-rhel7-ds.xml 
Connecting to 'rhel7.6-kicked' on port '22'...
Connected!
Copying input file './ssg-rhel7-ds.xml' to remote working directory '/tmp/tmp.bMdBnKwcR6'...
ssg-rhel7-ds.xml                                                                                                                                                                 100%   26MB 163.4MB/s   00:00    
Starting the evaluation...
WARNING: Datastream component 'scap_org.open-scap_cref_security-data-oval-com.redhat.rhsa-RHEL7.xml.bz2' points out to the remote 'https://www.redhat.com/security/data/oval/com.redhat.rhsa-RHEL7.xml.bz2'. Use '--fetch-remote-resources' option to download it.
WARNING: Skipping 'https://www.redhat.com/security/data/oval/com.redhat.rhsa-RHEL7.xml.bz2' file which is referenced from datastream
OpenSCAP Error: Could not extract scap_org.open-scap_cref_ssg-rhel7-xccdf-1.2.xml with all dependencies from datastream. [ds_sds_session.c:211]
oscap exit code: 1
Copying back requested files...
Removing remote temporary directory...
Disconnecting ssh and removing master ssh socket directory...
Exit request sent.
```

This should be fixed in OpenSCAP.

**Comment by jan-cerny** [comment]
@yuumasato Moreover, 
`oscap xccdf eval --fetch-remote-resources --profile ospp build/ssg-rhel8-ds.xml` fails because https://www.redhat.com/security/data/oval/com.redhat.rhsa-RHEL8.xml.bz2 doesn't exist yet, so that probably needs to be checked as well.

**Comment by jan-cerny** [comment]
Great job!

**Pull request merged**